src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -26,30 +57,104 @@ public class HttpServerImpl implements Server {
 
     /**
      * Constructor.
-     * @param port Http server port
+     * @param port HTTP server port
      * @param credentialRequestHandler request handler for server requests
      * @param executorService executor service instance
      * @throws IOException When server creation fails
      */
     HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService) throws IOException {
-        httpImpl = HttpServer.create(new InetSocketAddress("localhost", port), 0);
         this.credentialRequestHandler = credentialRequestHandler;
         this.executorService = executorService;
+        this.configuredPort = port;
+        this.servers = createServers(configuredPort);
+        this.resolvedPort = this.servers.get(0).getAddress().getPort();
+    }
+
+    /**
+     * Create ipv4 and ipv6 servers, both using the specified port.
+     *
+     * @param port chosen port for all created servers
+     * @return list of http servers
+     * @throws IOException if an I/O error occurs during server creation
+     * @throws BindException if the chosen port is unavailable
+     */
+    private List<HttpServer> createServers(int port) throws IOException {
+        HttpServer serverIPv4 = null;
+        InetSocketAddress addrIPv4 = new InetSocketAddress(ADDR_IPV4, port);
+        try {
+            serverIPv4 = HttpServer.create(addrIPv4, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV4)) {
+                throw e;
```

Comment:
what happens if we throw here? Will it retry? How do we recover?

Reply:
The thought was to let this bubble up as component ERROR and have the component retry.  In the case where port is 0, it's a good chance that the second time would pass.  (Side question--is there ever a time configured port is not the default of 0?  I know we have a `configuration.port` option defined in TokenExchangeService, but it doesn't appear to be set anywhere?)

If error persists and we get into a BROKEN state, then I imagine recovery would be a redeployment, the customer would need to clear up ports on their device

Reply:
The port may be set by the customer.

I think we'd want to retry indefinitely if the port is 0 because we should always be able to find some port to use.

Reply:
Sorry the above applies to the similar throw below for ipv6.  When this line throws it should only be possible if configured port != 0 and is already bound, so something is wrong with the configuration or the port needs to be unblocked (so i guess same question still stands if that port configuration is ever expected to change)

Reply:
"should only" since there's a bit of a race---in theory `HttpServer.create(addrIPv4, 0);` could throw when port == 0 when all ports are taken, then `canBindToAddress(ADDR_IPV4)` might return true if a port was freed between those two calls 🤷 

regardless, will add in explicit retrying when port is system-picked and bubble up exception when port is user configured 👍